dash/include/dash/halo/HaloMatrixWrapper.h

```diff
@@ -322,36 +375,64 @@ class HaloMatrixWrapper {
     }
   }
 
-  Element_t* halo_element_at(ElementCoords_t coords) {
+  /**
+   * Returns the halo value for a given global coordinate or nullptr if no halo
+   * element exists. This also means that only a unit connected to the given
+   * coordinate will return a halo value. All others will return nullptr.
+   */
+  Element_t* halo_element_at_global(ElementCoords_t coords) {
     const auto& offsets = _view_global.offsets();
     for(auto d = 0; d < NumDimensions; ++d) {
       coords[d] -= offsets[d];
     }
-    auto index       = _haloblock.index_at(_view_local, coords);
-    auto spec        = _halo_reg_spec.spec(index);
-    auto halomem_pos = _halomemory.pos_at(index);
+    auto        index       = _haloblock.index_at(_view_local, coords);
+    const auto& spec        = _halo_spec.spec(index);
+    auto        halomem_pos = _halomemory.pos_at(index);
+    if(spec.level() == 0 || halomem_pos == nullptr)
+      return nullptr;
+
+    if(!_halomemory.to_halo_mem_coords_check(index, coords))
+      return nullptr;
+
+    return _halomemory.pos_at(index) + _halomemory.offset(index, coords);
+  }
+
+  /**
+   * Returns the halo value for a given local coordinate or nullptr if no halo
+   * element exists.
+   */
+  Element_t* halo_element_at_local(ElementCoords_t coords) {
+    auto        index       = _haloblock.index_at(_view_local, coords);
+    const auto& spec        = _halo_spec.spec(index);
+    auto        halomem_pos = _halomemory.pos_at(index);
     if(spec.level() == 0 || halomem_pos == nullptr)
       return nullptr;
 
     if(!_halomemory.to_halo_mem_coords_check(index, coords))
       return nullptr;
 
-    return _halomemory.pos_at(index) + _halomemory.value_at(index, coords);
+    return _halomemory.pos_at(index) + _halomemory.offset(index, coords);
   }
 
-  template <typename IteratorT, typename FunctionT>
-  void compute(IteratorT inputBegin, IteratorT inputEnd, IteratorT outputBegin,
-               FunctionT func) {
-    while(inputBegin != inputEnd) {
-      *outputBegin = func(inputBegin);
-      ++inputBegin;
-      ++outputBegin;
+  /**
+   * Crates \ref HaloStencilOperator for a given \ref StencilSpec.
+   * Asserts whether the StencilSpec fits in the provided halo regions.
+   */
+  template <typename StencilSpecT>
+  HaloStencilOperator<Element_t, Pattern_t, StencilSpecT> stencil_operator(
+    const StencilSpecT& stencil_spec) {
+    for(const auto& stencil : stencil_spec.specs()) {
+      assert(stencil.max()
```

Comment:
Use `DASH_ASSERT[_MSG]` instead of `assert`

Reply:
done